### DISCUSSION TITLE
Added CUDA support

### DISCUSSION THREAD

**Discussion by 77axel** [77AXEL/PyCNN#13] [Announcements]
## 📦 Release: `v0.2.0 – Added support for CUDA`

This update enhances **PyCNN** by adding **CUDA (GPU) support**, allowing accelerated training and inference when a compatible GPU and CuPy are available. Users can now toggle between **CPU and GPU backends** seamlessly.

---

### 🚀 Key Features

* ✅ Fully functional CNN implementation from scratch
* 🧠 Manual convolution, max pooling, and ReLU activations
* 🔁 Forward and backward propagation with mini-batch gradient descent
* 🏷 Multi-class classification via softmax and cross-entropy loss
* 💾 Model save/load using `pickle`
* 🖼 RGB image preprocessing with customizable filters
* 🔍 Predict function to classify new unseen images
* 📊 Real-time training visualization (accuracy & loss per epoch)
* ⚡ **New:** Optional CUDA acceleration for faster training and inference
* 🔄 Automatic backend conversion when loading models trained on a different backend

---

#### Enable CUDA (GPU) support:

```python
from pycnn.model import CNN

model = CNN()
model.cuda(True)  # Enable CUDA
```

#### Switch back to CPU:

```python
model.cuda(False)  # Disable CUDA
```

> The model will automatically convert weights, biases, and datasets to the selected backend. Models trained on GPU can still be loaded on CPU and vice versa.

---

### 🏁 Training & Prediction

Training and prediction remain the same as previous versions. Example:

```python
model.init(
    image_size=32,
    batch_size=32,
    h1=128,
    h2=64,
    learning_rate=0.01,
    epochs=10,
    dataset_path="data",
    max_image=200
)
model.load_dataset()
model.train_model(visualize=True)
model.save_model()

model.load_model("model.bin")
result = model.predict("path/to/image.png")
print("Prediction:", result)
```

---

#### **v0.2.0**

* **New:** CUDA backend support via CuPy
* Automatic conversion between CPU and GPU backends
* Models can be trained on one backend and loaded on another seamlessly
* Minor improvements in training stability and performance

#### **v0.1.1**

* Real-time training visualization with Matplotlib

#### **v0.1.0**

* Initial version with full training and prediction pipeline

---

### 📌 Installation

```bash
pip install git+https://github.com/77AXEL/PyCNN.git@v0.2.0
```

> Optional: Install CuPy for CUDA support:

```bash
pip install cupy-cuda118  # Match your CUDA version
```

---

### 💬 Feedback & Contributions

We welcome issues, suggestions, and contributions!
Check the [Discussions tab](https://github.com/77AXEL/PyCNN/discussions) or see [CONTRIBUTING.md](https://chatgpt.com/c/CONTRIBUTING.md)

---

### 🛡 Security

Found a security issue? Please report privately to:
📧 [a.x.e.l777444000@gmail.com](mailto:a.x.e.l777444000@gmail.com)

---

### 📜 License

Released under the <a href="https://github.com/77AXEL/PyCNN/blob/b7becb4bef3b0156dad9397b1d95d6465e98fc3c/LICENSE">MIT License</a>

<hr /><em>This discussion was created from the release <a href='https://github.com/77AXEL/PyCNN/releases/tag/v0.1.2'>Added CUDA support</a>.</em>
